subsys/net/lib/coap/coap.c

```diff
@@ -1817,23 +1819,44 @@ struct coap_reply *coap_response_received(
 		return NULL;
 	}
 
+	type = coap_header_get_type(response);
 	id = coap_header_get_id(response);
 	tkl = coap_header_get_token(response, token);
 
+	if (type == COAP_TYPE_ACK || type == COAP_TYPE_RESET) {
+		piggybacked = true;
+	}
+
 	for (i = 0, r = replies; i < len; i++, r++) {
 		int age;
 
 		if ((r->id == 0U) && (r->tkl == 0U)) {
 			continue;
 		}
 
-		/* Piggybacked must match id when token is empty */
-		if ((r->id != id) && (tkl == 0U)) {
-			continue;
-		}
+		if (piggybacked) {
+			/* In a piggybacked response, the Message ID of the
+			 * Confirmable request and the Acknowledgment MUST
+			 * match, and the tokens of the response and original
+			 * request MUST match.
+			 */
+			if ((r->id != id || (r->tkl != tkl))) {
+				continue;
+			}
 
-		if (tkl > 0 && memcmp(r->token, token, tkl)) {
-			continue;
+			if (r->tkl > 0) {
+				if (memcmp(r->token, token, r->tkl) != 0) {
+					continue;
+				}
+			}
```

Comment:
```
if (piggybacked) {
   if ((r->id != id) {
      continue;
   }   
}

if ((r->tkl != tkl) {
   continue;
}   

if (r->tkl > 0) {
   if (memcmp(r->token, token, r->tkl) != 0) {
      continue;
   }
}
```

will use less code-duplicates.

Reply:
Thanks for the suggestion, I took the wrong assumption that for a separate response matching token means non-empty, which apparently is not the case. Simplified the code as suggested, and updated relevant test cases.